pkg/operator/controller/ingress/router_status.go

```diff
@@ -0,0 +1,212 @@
+		return fmt.Errorf("failed to list all routes in order to clear route status")
+	}
+
+	// List namespaces filtered by our ingress's namespace selector.
+	namespaceSelector, err := metav1.LabelSelectorAsSelector(ingress.Spec.NamespaceSelector)
+	if err != nil {
+		return fmt.Errorf("ingresscontroller %s has an invalid namespace selector: %w", ingress.Name, err)
+	}
+	namespaceLabelSelector, _ := labels.Parse(namespaceSelector.String())
+	filteredNamespaceList := &corev1.NamespaceList{}
+	if err := r.cache.List(context.TODO(), filteredNamespaceList,
+		client.MatchingLabelsSelector{Selector: namespaceLabelSelector}); err != nil {
+		return fmt.Errorf("failed to list all routes in order to clear route status for %s", ingress.Name)
+	}
+
+	// Create a set of namespaces to easily look up namespaces in this shard.
+	namespacesInShard := sets.NewString()
+	for i := range filteredNamespaceList.Items {
+		namespacesInShard.Insert(filteredNamespaceList.Items[i].Name)
+	}
+
+	// List routes filtered by our ingress's route selector.
+	routeSelector, err := metav1.LabelSelectorAsSelector(ingress.Spec.RouteSelector)
+	if err != nil {
+		return fmt.Errorf("ingresscontroller %s has an invalid route selector: %w", ingress.Name, err)
+	}
+	routeLabelSelector, _ := labels.Parse(routeSelector.String())
+	filteredRouteList := &routev1.RouteList{}
+	if err := r.cache.List(context.TODO(), filteredRouteList,
+		client.MatchingLabelsSelector{Selector: routeLabelSelector}); err != nil {
+		return fmt.Errorf("failed to list all routes in order to clear route status")
+	}
+
+	// Create a map of routes to easily look up routes in this shard.
+	type routeKey struct {
+		name, namespace string
+	}
+	routesInShard := map[routeKey]struct{}{}
+	for i := range filteredRouteList.Items {
+		route := &filteredRouteList.Items[i]
+		routesInShard[routeKey{route.Name, route.Namespace}] = struct{}{}
+	}
+
+	// Iterate over the entire route list and clear if not selected by route selector OR namespace selector.
+	routesCleared := 0
+	for i := range routeList.Items {
+		route := &routeList.Items[i]
+		_, routeInShard := routesInShard[routeKey{route.Name, route.Namespace}]
+		namespaceInShard := namespacesInShard.Has(route.Namespace)
+
+		if !routeInShard || !namespaceInShard {
+			cleared, _ := r.clearRouteStatus(route, ingress.ObjectMeta.Name)
+			if cleared {
+				log.Info("cleared route from ingress controller", "Route", route.Namespace+"/"+route.Name, "Ingress Controller", ingress.Name)
+				routesCleared++
+			}
+		}
+
+	}
+	elapsed := time.Since(start)
+	log.Info("cleared route status after selector update", "Ingress Controller", ingress.Name, "Routes Status Cleared", routesCleared, "Time Elapsed", elapsed)
+	return nil
+}
+
+// findCondition locates the first condition that corresponds to the requested type.
+func findCondition(ingress *routev1.RouteIngress, t routev1.RouteIngressConditionType) (_ *routev1.RouteIngressCondition) {
```

Comment:
Kind of weird to use the syntax for named results and then use the blank identifier.  * grin *.  
```suggestion
func findCondition(ingress *routev1.RouteIngress, t routev1.RouteIngressConditionType) *routev1.RouteIngressCondition {
```

Reply:
...I have to admit something...I stole this code from: https://github.com/openshift/router/blob/88ecceb8953751fcde2fe2eab52a5c15e07a955e/pkg/router/controller/status.go#L250

But yes this is silly. Done.